move quantization normalization layers to aten/src/ATen/native/quantized/cpu/

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37802 quant: clean up normalization channels_last handling
* #37781 quant: add test util for testing off by one quantized tensors
* **#37352 move quantization normalization layers to aten/src/ATen/native/quantized/cpu/**
* #36847 add quantized instancenorm operator
* #36835 add quantized groupnorm operator
* #37947 quant: remove hypothesis and int32 from layernorm test

Summary:

Implementing cleanup requested on #36835.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_instance_norm
python test/test_quantization.py TestQuantizedOps.test_group_norm
python test/test_quantization.py TestQuantizedOps.test_qlayer_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21261139](https://our.internmc.facebook.com/intern/diff/D21261139)

vkuzo committed:
move quantization normalization layers to aten/src/ATen/native/quantized/cpu/

Summary:

Implementing cleanup requested on #36835.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_instance_norm
python test/test_quantization.py TestQuantizedOps.test_group_norm
python test/test_quantization.py TestQuantizedOps.test_qlayer_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on aten/src/ATen/native/layer_norm.h:
```diff
@@ -6,6 +6,72 @@
+      1LL,
+      std::multiplies<int64_t>());
+
+  const auto& X = input.is_contiguous() ? input : input.contiguous();
```
What happens when the input is in NHWC (as is the case after a quantized op)? Can we avoid the call to contiguous here?

vkuzo replied:
we do need the input in NCHW for performant inference, for group/layer/instance norm.  Let me follow up offline if there are any changes in tests or output format that we might need.  If it's ok, let me do any needed changes in a future PR, since this PR is a refactor with no logic changes

raghuramank100 commented on aten/src/ATen/native/quantized/cpu/quantized_ops.h:
```diff
@@ -118,6 +118,18 @@ using qtopk_fn = void(*)(Tensor&, Tensor&, const Tensor&, int64_t, int64_t, bool
 
 using qbatch_norm_fn = void(*)(int64_t, int64_t, int64_t, int64_t, int64_t, const Tensor&, const Tensor&, const Tensor&, Tensor&);
 
+using forward_quantized_fn = void (*)(
```
Can we have a name that is more representative like qlayer_norm_fn?

raghuramank100 reviewed (changes requested):
Looks good, but had a question on calls to contiguous that can impact perf.

vkuzo committed:
Update on "move quantization normalization layers to aten/src/ATen/native/quantized/cpu/"


Summary:

Implementing cleanup requested on #36835.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_instance_norm
python test/test_quantization.py TestQuantizedOps.test_group_norm
python test/test_quantization.py TestQuantizedOps.test_qlayer_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21261139](https://our.internmc.facebook.com/intern/diff/D21261139)

[ghstack-poisoned]

vkuzo committed:
Update on "move quantization normalization layers to aten/src/ATen/native/quantized/cpu/"


Summary:

Implementing cleanup requested on #36835.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_instance_norm
python test/test_quantization.py TestQuantizedOps.test_group_norm
python test/test_quantization.py TestQuantizedOps.test_qlayer_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21261139](https://our.internmc.facebook.com/intern/diff/D21261139)

[ghstack-poisoned]

vkuzo committed:
Update on "move quantization normalization layers to aten/src/ATen/native/quantized/cpu/"


Summary:

Implementing cleanup requested on #36835.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_instance_norm
python test/test_quantization.py TestQuantizedOps.test_group_norm
python test/test_quantization.py TestQuantizedOps.test_qlayer_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21261139](https://our.internmc.facebook.com/intern/diff/D21261139)

[ghstack-poisoned]

vkuzo committed:
Update on "move quantization normalization layers to aten/src/ATen/native/quantized/cpu/"


Summary:

Implementing cleanup requested on #36835.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_instance_norm
python test/test_quantization.py TestQuantizedOps.test_group_norm
python test/test_quantization.py TestQuantizedOps.test_qlayer_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21261139](https://our.internmc.facebook.com/intern/diff/D21261139)

[ghstack-poisoned]
